src/main/java/com/squareup/kotlinpoet/AnnotationSpec.kt

```diff
@@ -30,6 +30,25 @@ class AnnotationSpec private constructor(builder: AnnotationSpec.Builder) {
   val className: ClassName = builder.className
   val members = builder.members.toImmutableList()
   val useSiteTarget: UseSiteTarget? = builder.useSiteTarget
+  private val tags: Map<KClass<*>, Any> = builder.tags.toImmutableMap()
+
+  /**
+   * Returns the tag attached with [Any] as a key, or `null` if no tag is attached with that key.
+   */
+  fun tag(): Any? = tag(Any::class)
+
+  /** Returns the tag attached with [type] as a key, or null if no tag is attached with that key. */
+  fun <T : Any> tag(type: Class<out T>): T? = tag(type.kotlin)
+
+  /** Returns the tag attached with [type] as a key, or null if no tag is attached with that key. */
+  fun <T : Any> tag(type: KClass<out T>): T? {
+    @Suppress("UNCHECKED_CAST")
+    return tags[type] as? T
```

Comment:
```suggestion
    return tags[type] as T?
```

Reply:
Same change on all specs

Reply:
Actually, `type.safeCast(tags[type])` is probably best here.

Reply:
Hm... I think your first suggestion (`tags[type] as T?`) is the right thing to do actually. I was trying to just make it null-safe but throw on type mismatches since they're programmer error. What do you think?

Reply:
`safeCast` is the safer version of that because it actually performs a cast at runtime instead of being erased into a no-op cast to `Object`.

Reply:
Right, but it will always fail the cast silently and just return null rather than throw it back to the caller. Is that what we want?

![image](https://user-images.githubusercontent.com/1361086/53691635-fb96c080-3d36-11e9-8105-543c295f3416.png)

Reply:
If we want to return null for missing keys then this is fine, otherwise we can do
```
type.safeCast(tags[type]) ?: throw NoSuchElementException("Missing tag for type $type")
```

Reply:
Wow safeCast does not do what I thought it did. And cast is kinda shitty. I guess I'm looking for `?.let(KClass::cast)` where there's an actual type check that tolerates null.